rootsanalyzer/rootsanalyzer.go

```diff
@@ -0,0 +1,152 @@
+				// TODO(RJPercival): If the root set is flapping back to what it recently was, suppress sending an incident report
+				// since it could just be the result of skew between log frontends. However, if it doesn't flap back to the new
+				// root set again within a certain amount of time, then the suppressed report should be sent.
+				oldRoots, err := st.ReadRoots(ctx, lastRootSetID)
+				if err != nil {
+					glog.Errorf("rootsanalyzer: %s", err)
+					return
+				}
+				newRoots, err := st.ReadRoots(ctx, rootSetID)
+				if err != nil {
+					glog.Errorf("rootsanalyzer: %s", err)
+					return
+				}
+				addedCerts, removedCerts := diffRootSets(oldRoots, newRoots)
+				if err := reportChange(ctx, rep, l, addedCerts, removedCerts); err != nil {
+					glog.Errorf("rootsanalyzer: %s", err)
+					return
+				}
+			}
+			lastRootSetID = rootSetID
+		}
+	}
+}
+
+// diffRootSets returns the certificates that have been added or removed in new, relative to old.
+// Neither old nor new are allowed to contain duplicates.
+func diffRootSets(old, new []*x509.Certificate) (added, removed []*x509.Certificate) {
+	oldSet := make(map[string]*x509.Certificate, len(old))
+	for _, cert := range old {
+		oldSet[string(cert.Raw)] = cert
+	}
+	// This algorithm assumes that there are no duplicates in new.
+	for _, cert := range new {
+		certDER := string(cert.Raw)
+		if oldSet[certDER] != nil {
+			// cert appears in both old and new - remove it from oldSet
+			// so that, after the loop, oldSet contains only certs
+			// that are in old but not new.
+			delete(oldSet, certDER)
+		} else {
+			// cert is only in new.
+			added = append(added, cert)
+		}
+	}
+	for _, cert := range oldSet {
+		removed = append(removed, cert)
+	}
+	return added, removed
+}
+
+// sortCerts sorts a slice of certificates first by their subject, then by their raw DER.
+func sortCerts(certs []*x509.Certificate) {
+	sort.Slice(certs, func(i, j int) bool {
+		if subj1, subj2 := certs[i].Subject.String(), certs[j].Subject.String(); subj1 != subj2 {
+			return subj1 < subj2
+		}
+		return bytes.Compare(certs[i].Raw, certs[j].Raw) < 0
+	})
+}
+
+func reportChange(ctx context.Context, rep incident.Reporter, l *ctlog.Log, addedCerts, removedCerts []*x509.Certificate) error {
+	fullURL := l.URL
+	getRootsURL, err := url.Parse(l.URL)
+	if err != nil {
+		glog.Errorf("Failed to parse CT Log URL %q: %v", l.URL, err)
+	} else {
+		getRootsURL.Path = path.Join(getRootsURL.Path, ct.GetRootsPath)
+		fullURL = getRootsURL.String()
+	}
+
+	// Sort certs so that the report is deterministic - makes testing easier.
+	sortCerts(addedCerts)
+	sortCerts(removedCerts)
+
+	var strBuilder strings.Builder
+	if err := incidentTemplate.Execute(&strBuilder, incidentTemplateArgs{
+		Log:          l,
+		AddedCerts:   addedCerts,
+		RemovedCerts: removedCerts,
+	}); err != nil {
+		return err
+	}
+	rep.Log(ctx, l.URL, "Root certificates changed", "roots", fullURL, strBuilder.String())
```

Comment:
(Musing) It's pretty surprising that Log and Logf in the incident.Reporter interface don't give the option to return and error.... Might need to tweak that in the future.  (Not for this PR of course!)